### ISSUE TITLE
mlflow model cannot be loaded

### ISSUE THREAD

**Issue by fschlz** [kserve/kserve#3733]
/kind bug

**What steps did you take and what happened:**

Hey guys, I am trying to use KServer on AKS.
I installed all the dependencies on AKS and am trying to deploy a test inference service.
However, the model isn't getting loaded correctly.
Locally everything works out right.
Unfortunately, the service doesn't seem to recognize the model files I have registered.
Plus, the environment that is created doesn't seem to respect the version numbers that are set in `requirements.txt`

```log
Environment tarball not found at '/mnt/models/environment.tar.gz'
Environment not found at './envs/environment'
2024-06-11 14:31:10,008 [mlserver.parallel] DEBUG - Starting response processing loop...
2024-06-11 14:31:10,009 [mlserver.rest] INFO - HTTP server running on http://0.0.0.0:8080
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2024-06-11 14:31:10,083 [mlserver.metrics] INFO - Metrics server running on http://0.0.0.0:8082
2024-06-11 14:31:10,083 [mlserver.metrics] INFO - Prometheus scraping endpoint can be accessed on http://0.0.0.0:8082/metrics
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
2024-06-11 14:31:11,102 [mlserver.grpc] INFO - gRPC server running on http://0.0.0.0:9000
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
INFO:     Uvicorn running on http://0.0.0.0:8082 (Press CTRL+C to quit)
2024/06/11 14:31:12 WARNING mlflow.pyfunc: Detected one or more mismatches between the model's dependencies and the current Python environment:
- mlflow (current: 2.3.1, required: mlflow==2.12.2)
- cloudpickle (current: 2.2.1, required: cloudpickle==3.0.0)
- numpy (current: 1.23.5, required: numpy==1.24.4)
- packaging (current: 23.1, required: packaging==23.2)
- psutil (current: uninstalled, required: psutil==5.9.8)
- pyyaml (current: 6.0, required: pyyaml==6.0.1)
- scikit-learn (current: 1.2.2, required: scikit-learn==1.3.2)
- scipy (current: 1.9.1, required: scipy==1.10.1)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.
2024-06-11 14:31:12,049 [mlserver] INFO - Couldn't load model 'wine-classifier'. Model will be removed from registry.
2024-06-11 14:31:12,049 [mlserver.parallel] ERROR - An error occurred processing a model update of type 'Load'.
Traceback (most recent call last):
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/worker.py", line 158, in _process_model_update
await self._model_registry.load(model_settings)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 293, in load
return await self._models[model_settings.name].load(model_settings)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 148, in load
await self._load_model(new_model)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 165, in _load_model
model.ready = await model.load()
File "/opt/conda/lib/python3.8/site-packages/mlserver_mlflow/runtime.py", line 155, in load
self._model = mlflow.pyfunc.load_model(model_uri)
File "/opt/conda/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 582, in load_model
model_meta = Model.load(os.path.join(local_path, MLMODEL_FILE_NAME))
File "/opt/conda/lib/python3.8/site-packages/mlflow/models/model.py", line 468, in load
return cls.from_dict(yaml.safe_load(f.read()))
File "/opt/conda/lib/python3.8/site-packages/mlflow/models/model.py", line 478, in from_dict
model_dict["signature"] = ModelSignature.from_dict(model_dict["signature"])
File "/opt/conda/lib/python3.8/site-packages/mlflow/models/signature.py", line 83, in from_dict
inputs = Schema.from_json(signature_dict["inputs"])
File "/opt/conda/lib/python3.8/site-packages/mlflow/types/schema.py", line 360, in from_json
return cls([read_input(x) for x in json.loads(json_str)])
File "/opt/conda/lib/python3.8/site-packages/mlflow/types/schema.py", line 360, in <listcomp>
return cls([read_input(x) for x in json.loads(json_str)])
File "/opt/conda/lib/python3.8/site-packages/mlflow/types/schema.py", line 358, in read_input
return TensorSpec.from_json_dict(**x) if x["type"] == "tensor" else ColSpec(**x)
TypeError: __init__() got an unexpected keyword argument 'required'
2024-06-11 14:31:12,051 [mlserver] INFO - Couldn't load model 'wine-classifier'. Model will be removed from registry.
2024-06-11 14:31:12,052 [mlserver.parallel] ERROR - An error occurred processing a model update of type 'Unload'.
Traceback (most recent call last):
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/worker.py", line 160, in _process_model_update
await self._model_registry.unload_version(
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 302, in unload_version
await model_registry.unload_version(version)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 201, in unload_version
model = await self.get_model(version)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 237, in get_model
raise ModelNotFound(self._name, version)
mlserver.errors.ModelNotFound: Model wine-classifier not found
2024-06-11 14:31:12,053 [mlserver] ERROR - Some of the models failed to load during startup!
Traceback (most recent call last):
File "/opt/conda/lib/python3.8/site-packages/mlserver/server.py", line 125, in start
await asyncio.gather(
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 293, in load
return await self._models[model_settings.name].load(model_settings)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 148, in load
await self._load_model(new_model)
File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 161, in _load_model
model = await callback(model)
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/registry.py", line 152, in load_model
loaded = await pool.load_model(model)
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/pool.py", line 74, in load_model
await self._dispatcher.dispatch_update(load_message)
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 123, in dispatch_update
return await asyncio.gather(
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 138, in _dispatch_update
return await self._dispatch(worker_update)
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 146, in _dispatch
return await self._wait_response(internal_id)
File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 152, in _wait_response
inference_response = await async_response
mlserver.parallel.errors.WorkerError: builtins.TypeError: __init__() got an unexpected keyword argument 'required'
2024-06-11 14:31:12,053 [mlserver.parallel] INFO - Waiting for shutdown of default inference pool...
2024-06-11 14:31:12,193 [mlserver.parallel] INFO - Shutdown of default inference pool complete
2024-06-11 14:31:12,193 [mlserver.grpc] INFO - Waiting for gRPC server shutdown
2024-06-11 14:31:12,196 [mlserver.grpc] INFO - gRPC server shutdown complete
INFO:     Shutting down
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
```

Does anyone know what could be wrong?

**What did you expect to happen:**

Deploy the inference service.

**What's the InferenceService yaml:**

```yaml
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "wine-classifier"
  namespace: "mlflow-kserve-test"
spec:
  predictor:
    serviceAccountName: sa-azure
    model:
      modelFormat:
        name: mlflow
      protocolVersion: v2
      storageUri: "https://{SA}.blob.core.windows.net/azureml/ExperimentRun/dcid.{RUN_ID}/model"
```

**Anything else you would like to add:**

These are the model files in my Storage Account (see screenshot)
![Image from kserve](https://github.com/kserve/kserve/assets/22344801/612c580c-8c3a-4580-9fa2-f35575338964)

**Environment:**

- Istio Version: 1.22.1
- Knative Version: 1.14.0
- KServe Version: 0.12.0
- Kubeflow version: -
- Cloud Environment: k8s_istio on Azure Kubernetes Service
- Kubernetes version: 1.28.5
- OS (e.g. from `/etc/os-release`): ubuntu 22.04

**Comment by 5uperpalo** [comment]
1. I came across the same issue when using mlflow custom model `mlflow.pyfunc.PythonModel`
2. the model works when I server it locally using `mlflow serve`
3. I tried to make a docker image using `mlflow models build-docker` - in this case the pod crashes as it is not responding to ReadyProbe
4. I tried to tar.gzip the model and upload it to github (same as using s3, i.e. same as the result from @fschlz )
5. I tried to create conda tar.gz environment file as in the exmaples, here: (i) https://mlserver.readthedocs.io/en/latest/examples/conda/README.html , (ii) https://docs.seldon.io/projects/seldon-core/en/latest/examples/mlflow_v2_protocol_end_to_end.html

using last approach, i.e. by creating custom conda env, by downloading conda.yaml from s3 and creating tar.gz:
```
conda env create --force -f conda.yaml
conda-pack --name mlflow-env --output environment.tar.gz --force --ignore-editable-packages --ignore-missing-files
```

I got a bit further but now I am stuck on `conda-unpack: error: unrecognized arguments: --quiet`

```
lib/python3.10/site-packages/sklearn/ensemble/_hist_gradient_boosting/tests/__pycache__/test_predictor.cpython-310.pyc
lib/python3.10/site-packages/matplotlib/__pycache__/offsetbox.cpython-310.pyc
lib/python3.10/site-packages/fsspec/implementations/__pycache__/reference.cpython-310.pyc
lib/python3.10/site-packages/mlflow/store/tracking/dbmodels/__pycache__/models.cpython-310.pyc
bin/activate
bin/deactivate
bin/conda-unpack
--> Sourcing new environment at ./envs/environment...
--> Calling conda-unpack...
usage: conda-unpack [-h] [--version]
conda-unpack: error: unrecognized arguments: --quiet
--> Disabling user-installed packages...
2024-06-13 15:28:27,004 [mlserver.parallel] DEBUG - Starting response processing loop...
2024-06-13 15:28:27,008 [mlserver.rest] INFO - HTTP server running on http://0.0.0.0:8080
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2024-06-13 15:28:27,105 [mlserver.metrics] INFO - Metrics server running on http://0.0.0.0:8082
2024-06-13 15:28:27,106 [mlserver.metrics] INFO - Prometheus scraping endpoint can be accessed on http://0.0.0.0:8082/metrics
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
2024-06-13 15:28:29,437 [mlserver.grpc] INFO - gRPC server running on http://0.0.0.0:9000
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
INFO:     Uvicorn running on http://0.0.0.0:8082 (Press CTRL+C to quit)
2024/06/13 15:28:30 WARNING mlflow.pyfunc: Detected one or more mismatches between the model's dependencies and the current Python environment:
 - mlflow (current: 2.3.1, required: mlflow==2.13.0)
 - cryptography (current: 38.0.4, required: cryptography==40.0.2)
 - cytoolz (current: uninstalled, required: cytoolz==0.12.0)
 - defusedxml (current: uninstalled, required: defusedxml==0.7.1)
 - distributed (current: uninstalled, required: distributed==2023.5.1)
 - lightgbm (current: 3.3.5, required: lightgbm==4.3.0)
 - lz4 (current: uninstalled, required: lz4==4.3.2)
 - numpy (current: 1.23.5, required: numpy==1.24.3)
 - optuna-integration (current: uninstalled, required: optuna-integration==3.6.0)
 - pandas (current: 2.0.1, required: pandas==2.2.2)
 - pyarrow (current: 11.0.0, required: pyarrow==12.0.0)
 - pynvml (current: uninstalled, required: pynvml==11.5.0)
 - redis (current: uninstalled, required: redis==5.0.4)
 - scipy (current: 1.9.1, required: scipy==1.10.1)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.
2024/06/13 15:28:30 WARNING mlflow.pyfunc: The version of Python that the model was saved in, `Python 3.10.11`, differs from the version of Python that is currently running, `Python 3.8.16`, and may be incompatible
2024-06-13 15:28:30,770 [mlserver] INFO - Couldn't load model 'success6g-model'. Model will be removed from registry.
2024-06-13 15:28:30,770 [mlserver.parallel] ERROR - An error occurred processing a model update of type 'Load'.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/worker.py", line 158, in _process_model_update
    await self._model_registry.load(model_settings)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 293, in load
    return await self._models[model_settings.name].load(model_settings)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 148, in load
    await self._load_model(new_model)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 165, in _load_model
    model.ready = await model.load()
  File "/opt/conda/lib/python3.8/site-packages/mlserver_mlflow/runtime.py", line 155, in load
    self._model = mlflow.pyfunc.load_model(model_uri)
  File "/opt/conda/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 596, in load_model
    model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "/opt/conda/lib/python3.8/site-packages/mlflow/pyfunc/model.py", line 291, in _load_pyfunc
    python_model = cloudpickle.load(f)
ModuleNotFoundError: No module named 'inference_model'
2024-06-13 15:28:30,773 [mlserver] INFO - Couldn't load model 'success6g-model'. Model will be removed from registry.
2024-06-13 15:28:30,778 [mlserver.parallel] ERROR - An error occurred processing a model update of type 'Unload'.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/worker.py", line 160, in _process_model_update
    await self._model_registry.unload_version(
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 302, in unload_version
    await model_registry.unload_version(version)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 201, in unload_version
    model = await self.get_model(version)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 237, in get_model
    raise ModelNotFound(self._name, version)
mlserver.errors.ModelNotFound: Model success6g-model not found
2024-06-13 15:28:30,779 [mlserver] ERROR - Some of the models failed to load during startup!
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/mlserver/server.py", line 125, in start
    await asyncio.gather(
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 293, in load
    return await self._models[model_settings.name].load(model_settings)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 148, in load
    await self._load_model(new_model)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/registry.py", line 161, in _load_model
    model = await callback(model)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/registry.py", line 152, in load_model
    loaded = await pool.load_model(model)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/pool.py", line 74, in load_model
    await self._dispatcher.dispatch_update(load_message)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 123, in dispatch_update
    return await asyncio.gather(
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 138, in _dispatch_update
    return await self._dispatch(worker_update)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 146, in _dispatch
    return await self._wait_response(internal_id)
  File "/opt/conda/lib/python3.8/site-packages/mlserver/parallel/dispatcher.py", line 152, in _wait_response
    inference_response = await async_response
mlserver.parallel.errors.WorkerError: builtins.ModuleNotFoundError: No module named 'inference_model'
2024-06-13 15:28:30,781 [mlserver.parallel] INFO - Waiting for shutdown of default inference pool...
2024-06-13 15:28:30,962 [mlserver.parallel] INFO - Shutdown of default inference pool complete
2024-06-13 15:28:30,962 [mlserver.grpc] INFO - Waiting for gRPC server shutdown
2024-06-13 15:28:30,969 [mlserver.grpc] INFO - gRPC server shutdown complete
INFO:     Shutting down
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
```

does anybody know how to fix this issue? I am able to serve basic model examples but not custom models on Kserve and I already spend week of my life on this ... my next step is to try SeldonCore, I guess...

**Comment by JoseeePZ** [comment]
@5uperpalo 

> does anybody know how to fix this issue?

Did your log your extra pip requirements in mlflow? The conda.yaml which is created while logging a model in mlflow only picks up on the flavor you used for logging

see [Dependencies in mlflow](https://mlflow.org/docs/latest/model/dependencies.html)

mlserver and mlserver-mlflow also need to be in the conda.yaml

**Comment by 5uperpalo** [comment]
@JoseeePZ I noticed the code dependency in mlflow, but the logic is hard to follow for me... if I understand it correctly this is not my case as I am not doing any imports inside the code and the `log_model` in mlflow correctly inferred also optuna that is part of the custom Trainer object ... this is part of the SUCCESS6G project - https://github.com/5uperpalo/success6g-edge/blob/main/notebooks/inference_model/04_model_training_pipeline.ipynb, so the code is publicly available (the passwords etc. in the code do not matter, as kubernetes is not publicly accessible) ... 

also, if I understand it correctly, if the "dependencies in mlflow" were an issue I would not be able to serve the model locally ... and as opposed to @fschlz I was "lucky" enough to be able to serve at least the wine classifier example model, as you might see in the log, if I want to use "custom conda env" with different python version and additional packages I am stuck on conde-unpack ... I am guessing that somewhere inside kserve there is some hidden legacy command to run all commands inside the container with `--quite` argument, and as conda-unpack does not have it - it fails?

**Comment by 5uperpalo** [comment]
again, I am just guessing... with newer versions of mlflow/mlflow-server?/kserve there is some issue that causes:
1. mlflow docker to not respond to probes when container is created by kserve
2. kserve ignoring requirements.txt when creating mlflow model container
3. kserve running conda-unpack with --quite argument

**Comment by 5uperpalo** [comment]
ok, I found the issue (by luck I guess) ... again, if I understand it correctly kserve uses MLserver, which runs the following script, where I got stuck on this line : https://github.com/SeldonIO/MLServer/blob/3af34a0c04893dabe24b1e416da2515f3c6e63c5/hack/activate-env.sh#L43 ... this line was changed in the master exactly year ago, so Kserve is using version of MLserver that is at least year old! : https://github.com/SeldonIO/MLServer/pull/1230/files .. this was also completely removed from cli/constants.py afterwards - https://github.com/SeldonIO/MLServer/blob/3af34a0c04893dabe24b1e416da2515f3c6e63c5/mlserver/cli/constants.py#L31 .. my best guess, if/when Kserve starts to use recent version MLserver version it will fix my and @fschlz issue

**Comment by jkoessle** [comment]
The latest release of kserve (v0.13) seems to use mlserver 1.3.2: [kserve-cluster-resources.yaml](https://github.com/kserve/kserve/blob/master/install/v0.13.0-rc0/kserve-cluster-resources.yaml#L80)
However, the fix is only available as of mlserver 1.3.4, where its also mentioned in the changelog: [Release 1.3.4](https://github.com/SeldonIO/MLServer/releases/tag/1.3.4)

**Comment by sivanantha321** [comment]
@5uperpalo @jkoessle Thanks for bringing this in. Will raise a PR to upgrade the MLServer shortly. For now, You can just update the [MLserver image version in ClusterServingRuntime CR](https://github.com/kserve/kserve/blob/212a77c2fd76255db49569d1efa12b4c075466a8/config/runtimes/kserve-mlserver.yaml#L35).

**Comment by fschlz** [comment]
Thanks for your input guys!
I updated the CRDs with [this](https://github.com/kserve/kserve/blob/master/install/v0.13.0-rc0/kserve-cluster-resources.yaml) manifest and using mlserver version 1.5.0.
mlserver still doesn't fully resolve the environment from the dependency files, but at least the service starts now and is operational.

Here's the output from the `kserve-container`
```
WARNING: No ICDs were found. Either,
- Install a conda package providing a OpenCL implementation (pocl, oclgrind, intel-compute-runtime, beignet) or
- Make your system-wide implementation visible by installing ocl-icd-system conda package.
Environment tarball not found at '/mnt/models/environment.tar.gz'
Environment not found at './envs/environment'
2024-06-14 13:22:09,911 [mlserver.parallel] DEBUG - Starting response processing loop...
2024-06-14 13:22:09,912 [mlserver.rest] INFO - HTTP server running on http://0.0.0.0:8080
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2024-06-14 13:22:09,926 [mlserver.metrics] INFO - Metrics server running on http://0.0.0.0:8082
2024-06-14 13:22:09,926 [mlserver.metrics] INFO - Prometheus scraping endpoint can be accessed on http://0.0.0.0:8082/metrics
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
2024-06-14 13:22:11,035 [mlserver.grpc] INFO - gRPC server running on http://0.0.0.0:9000
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
INFO:     Uvicorn running on http://0.0.0.0:8082 (Press CTRL+C to quit)
2024/06/14 13:22:12 WARNING mlflow.utils.requirements_utils: Detected one or more mismatches between the model's dependencies and the current Python environment:
- mlflow (current: 2.10.2, required: mlflow==2.13.2)
- psutil (current: uninstalled, required: psutil==5.9.8)
- scikit-learn (current: 1.4.1.post1, required: scikit-learn==1.5.0)
- scipy (current: 1.12.0, required: scipy==1.13.1)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.
/opt/conda/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator ElasticNet from version 1.5.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
warnings.warn(
2024-06-14 13:22:12,570 [mlserver][wine-classifier] INFO - Loaded model 'wine-classifier' successfully.
2024-06-14 13:22:12,573 [mlserver][wine-classifier] INFO - Loaded model 'wine-classifier' successfully.
INFO:     127.0.0.1:46322 - "POST /v2/models/wine-classifier/infer HTTP/1.1" 200 OK
`
``
